### DISCUSSION TITLE
Contents" 中键值对顺序问题

### DISCUSSION THREAD

**Discussion by polaris-gxf** [alibaba/ilogtail#409] [Help] [answered]
Contents":[{"Key":"_time_","Value":"2022-09-19T06:07:03.013684742Z"},{"Key":"_pod_name_","Value":"spartan-govern-manage-6ff6c8755d-9829d"},{"Key":"_namespace_","Value":"spartan"},{"Key":"_pod_uid_","Value":"2fce87be-899c-4f7a-bfcc-9f4e0516fe3f"},{"Key":"_container_ip_","Value":"192.168.46.78"},{"Key":"_container_name_","Value":"spartan-govern-manage"},{"Key":"content","Value":"#######"}]}

实践中发现 ，Contents" 中键值对的顺序输出的位置并不是固定不变的，请教这个顺序有什么办法固定下来么？

**Comment by messixukejia** [comment]
采集配置麻烦发下看看

**Reply by polaris-gxf** [reply]
inputs:
      - Type: service_docker_stdout
        Stdout: true
        Stderr: false
        IncludeLabel:
          annotation.io.kubernetes.container.ports: "[{\"containerPort\":8080,\"protocol\":\"TCP\"}]"
        ExcludeLabel:
          io.kubernetes.pod.namespace: "spartan-pre"
    processors:
      - Type: processor_split_log_regex
        SplitRegex: ^\[|^[0-9]{4}-[0-9]{2}-[0-9]{2}|^[0-9]{1,3}\.[0-9]{1,3}
        SplitKey: content
        PreserveOthers: true
      - Type: processor_drop
        DropKeys:
          - _image_name_
          - _source_
    flushers:
      - Type: flusher_kafka
        Brokers:
          - 192.168.0.166:9092
        Topic: spartan_java_log

**Reply by henryzhx8** [reply]
目前tag字段都是先存放在map里，再放到contents里的，所以暂时没办法保序。

**Reply by messixukejia** [reply]
@polaris-gxf 需求的背景能否描述下？为什么需要有保序的要求？

**Reply by polaris-gxf** [reply]
content中的kv能不能提升一个嵌套层次，放到contents里面。这样后面好处理，目前这样嵌套的太深了，量大了给后面的logstash很大压力。或者还有什么能在ilogtail前面处理的好办法？

**Reply by messixukejia** [reply]
可以关注下https://github.com/alibaba/ilogtail/discussions/280

**Reply by polaris-gxf** [reply]
如果不保序，后面需要在logstash中多加ruby语法处理，才能形成易分析的结构
ruby {
        code => "
        markets = event.get('Contents')
        markets.each_index do |index|
            event.set(markets[index]['Key'], markets[index]['Value'])
        end
    "
    }
====================================
{
      "_container_ip_" => "[192.168.18.240](http://192.168.18.240/)",
         "_namespace_" => "spartan",
          "_pod_name_" => "spartan-govern-ethsync-98c9d9d7-jrjn4",
            "sequence" => 0,
    "_container_name_" => "spartan-govern-ethsync",
              "_time_" => "2022-09-19T06:27:56.010841754Z",
                "Time" => 1663568876,
             "content" => "2022-09-19 06:27:56.010 INFO [TID: N/A][Thread-30] c.spartan.govern.ethsync.job.BlockTask: 区块采集------开始------",

**Reply by messixukejia** [reply]
https://github.com/alibaba/ilogtail/discussions/388

本周三会有双周会，会同步下kafka flusher的进展。会支持你预期的格式。

**Reply by polaris-gxf** [reply]
#280 是实现了，还是没实现呢？
我加了FlusherMode: 1       为什么没生效呢？

**Reply by messixukejia** [reply]
没实现，这个只是讨论。 具体进展双周会关注。

**Reply by polaris-gxf** [reply]
大约什么时候能用上？嘿嘿嘿！

**Comment by messixukejia** [comment, marked as answer]
https://github.com/alibaba/ilogtail/releases/tag/v1.3.0
https://github.com/alibaba/ilogtail/blob/main/docs/cn/data-pipeline/flusher/kafka_v2.md

kafkav2已发布，可以满足格式要求。

**Reply by polaris-gxf** [reply]
没看太懂，您指的是用processor_regex这个方案么？

**Reply by polaris-gxf** [reply]
理解有误，已经实现，谢谢。
